docs/prompting/self_criticism/self_refine.md

```diff
@@ -1,7 +1,214 @@
+    B --> C{Stopping<br>condition<br>met?}:::orange
+    C -->|No| D[Refine response]:::orange
+    C -->|Yes| E[Final output]:::green
+    D --> B
+    
+    classDef blue fill:#E3F2FD,stroke:#90CAF9,color:#1565C0
+    classDef orange fill:#FFF3E0,stroke:#FFE0B2,color:#E65100
+    classDef green fill:#E8F5E9,stroke:#A5D6A7,color:#2E7D32
+    linkStyle default stroke:#90A4AE,stroke-width:2px;
+    linkStyle 1,2,4 stroke:#FFB74D,stroke-width:2px;
+``` 
+
+```python hl_lines="88-92"
+import instructor
+from pydantic import BaseModel, Field
+from openai import OpenAI
+from typing import Optional
+
+
+class Response(BaseModel):
+    code: str
+
+
+class Feedback(BaseModel):
+    feedback: list[str] = Field(
+        description="A list of actions to take to improve the code."
+    )
+    done: bool
+
+
+class Timestep(BaseModel):
+    response: str
+    feedback: Optional[list[str]] = []
+    refined_response: Optional[str] = ""
+
+
+class History(BaseModel):
+    history: list[Timestep] = []
+
+
+client = instructor.from_openai(OpenAI())
+
+
+def generate_feedback(response):
+    return client.chat.completions.create(
+        model="gpt-4o",
+        response_model=Feedback,
+        messages=[
+            {
+                "role": "user",
+                "content": f"""
+                        You are an expert Python coder.
+                        Provide feedback on this code.
+                        How can we make it (1) faster and (2) more readable?
+                        {response.code}
+                        If the code does not need to be improved, then indicate by setting "done" to True.
+                        """,
+            }
+        ],
+    )
+
+
+def refine(response, feedback):
+    return client.chat.completions.create(
+        model="gpt-4o",
+        response_model=Response,
+        messages=[
+            {
+                "role": "user",
+                "content": f"""
+                        You are an expert Python coder.
+                        Given this response:
+                        {response.code}
+                        And this feedback:
+                        {feedback.feedback}
+                        Refine your response.
+                        """,
+            }
+        ],
+    )
+
+
+def stop_condition(feedback, history):
+    return feedback.done or len(history.history) >= 3
+
+
+if __name__ == "__main__":
+    response = client.chat.completions.create(
+        model="gpt-4o",
+        response_model=Response,
+        messages=[
+            {
+                "role": "user",
+                "content": "Write Python code to calculate the fibonacci sequence.",
+            }
+        ],
+    )
+
+    history = History()
+
+    while True:
+        feedback = generate_feedback(response)
+        if stop_condition(feedback, history):
+            break
+        refined_response = refine(response, feedback)
+
+        # Save to history
+        history.history.append(
```

Comment:
you can add a `def add(self, timestep)` method to the history class rather than doing history.history.append